profiling/cmd/catapult.go

```diff
@@ -0,0 +1,391 @@
+		return ti.BeginNsec < tj.BeginNsec
+	}
+	return ti.BeginSec < tj.BeginSec
+}
+
+func streamStatsCatapultJSON(s *snapshot, streamStatsCatapultJSONFileName string) error {
+	grpclog.Infof("calculating stream stats filters")
+	filterArray := strings.Split(*flagStreamStatsFilter, ",")
+	filter := make(map[string]bool)
+	for _, f := range filterArray {
+		filter[f] = true
+	}
+
+	grpclog.Infof("filter stream stats for %s", *flagStreamStatsFilter)
+	streamStats := make([]*ppb.Stat, 0)
+	for _, stat := range s.StreamStats {
+		if _, ok := filter[stat.Tags]; ok {
+			streamStats = append(streamStats, stat)
+		}
+	}
+
+	grpclog.Infof("sorting timers within all stats")
+	for id := range streamStats {
+		sort.Slice(streamStats[id].Timers, func(i, j int) bool {
+			return timerBeginIsBefore(streamStats[id].Timers[i], streamStats[id].Timers[j])
+		})
+	}
+
+	grpclog.Infof("sorting stream stats")
+	sort.Slice(streamStats, func(i, j int) bool {
+		if len(streamStats[j].Timers) == 0 {
+			return true
+		} else if len(streamStats[i].Timers) == 0 {
+			return false
+		}
+		pi := binary.BigEndian.Uint64(streamStats[i].Metadata[0:8])
+		pj := binary.BigEndian.Uint64(streamStats[j].Metadata[0:8])
+		if pi == pj {
+			return timerBeginIsBefore(streamStats[i].Timers[0], streamStats[j].Timers[0])
+		}
+
+		return pi < pj
+	})
+
+	// Clip the last stat as it's from the /Profiling/GetStreamStats call that we
+	// made to retrieve the stats themselves. This likely happened millions of
+	// nanoseconds after the last stream we want to profile, so it'd just make
+	// the catapult graph less readable.
+	if len(streamStats) > 0 {
+		streamStats = streamStats[:len(streamStats)-1]
+	}
+
+	// All timestamps use the earliest timestamp available as the reference.
+	grpclog.Infof("calculating the earliest timestamp across all timers")
+	var base *ppb.Timer
+	for _, stat := range streamStats {
+		for _, timer := range stat.Timers {
+			if base == nil || timerBeginIsBefore(base, timer) {
+				base = timer
+			}
+		}
+	}
+
+	grpclog.Infof("converting %d stats to catapult JSON format", len(streamStats))
+	jsonNodes := make([]jsonNode, 0)
+	for _, stat := range streamStats {
+		jsonNodes = append(jsonNodes, streamStatsCatapultJSONSingle(stat, base.BeginSec, base.BeginNsec)...)
+	}
+
+	grpclog.Infof("marshalling catapult JSON")
+	b, err := json.Marshal(jsonNodes)
+	if err != nil {
+		grpclog.Errorf("cannot marshal JSON: %v", err)
+		return err
+	}
+
+	grpclog.Infof("creating catapult JSON file")
+	streamStatsCatapultJSONFile, err := os.Create(streamStatsCatapultJSONFileName)
+	if err != nil {
+		grpclog.Errorf("cannot create file %s: %v", streamStatsCatapultJSONFileName, err)
+		return err
+	}
+
+	grpclog.Infof("writing catapult JSON to disk")
+	_, err = streamStatsCatapultJSONFile.Write(b)
+	if err != nil {
+		grpclog.Errorf("cannot write marshalled JSON: %v", err)
+		return err
+	}
+
+	grpclog.Infof("closing catapult JSON file")
+	streamStatsCatapultJSONFile.Close()
+	if err != nil {
```

Comment:
Did you forget to keep close's return error?

Reply:
Yep. Fixed.